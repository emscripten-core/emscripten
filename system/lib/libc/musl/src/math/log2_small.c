/* origin: FreeBSD /usr/src/lib/msun/src/e_log2.c */
/*
 * ====================================================
 * Copyright (C) 1993 by Sun Microsystems, Inc. All rights reserved.
 *
 * Developed at SunSoft, a Sun Microsystems, Inc. business.
 * Permission to use, copy, modify, and distribute this
 * software is freely granted, provided that this notice
 * is preserved.
 * ====================================================
 */
/*
 * Return the base 2 logarithm of x.  See log.c for most comments.
 *
 * Reduce x to 2^k (1+f) and calculate r = log(1+f) - f + f*f/2
 * as in log.c, then combine and scale in extra precision:
 *    log2(x) = (f - f*f/2 + r)/log(2) + k
 */

#include <math.h>
#include <stdint.h>

static const double
ivln2hi = 1.44269504072144627571e+00, /* 0x3ff71547, 0x65200000 */
ivln2lo = 1.67517131648865118353e-10, /* 0x3de705fc, 0x2eefa200 */
Lg1 = 6.666666666666735130e-01,  /* 3FE55555 55555593 */
Lg2 = 3.999999999940941908e-01,  /* 3FD99999 9997FA04 */
Lg3 = 2.857142874366239149e-01,  /* 3FD24924 94229359 */
Lg4 = 2.222219843214978396e-01,  /* 3FCC71C5 1D8E78AF */
Lg5 = 1.818357216161805012e-01,  /* 3FC74664 96CB03DE */
Lg6 = 1.531383769920937332e-01,  /* 3FC39A09 D078C69F */
Lg7 = 1.479819860511658591e-01;  /* 3FC2F112 DF3E5244 */

double log2(double x)
{
	union {double f; uint64_t i;} u = {x};
	double_t hfsq,f,s,z,R,w,t1,t2,y,hi,lo,val_hi,val_lo;
	uint32_t hx;
	int k;

	hx = u.i>>32;
	k = 0;
	if (hx < 0x00100000 || hx>>31) {
		if (u.i<<1 == 0)
			return -1/(x*x);  /* log(+-0)=-inf */
		if (hx>>31)
			return (x-x)/0.0; /* log(-#) = NaN */
		/* subnormal number, scale x up */
		k -= 54;
		x *= 0x1p54;
		u.f = x;
		hx = u.i>>32;
	} else if (hx >= 0x7ff00000) {
		return x;
	} else if (hx == 0x3ff00000 && u.i<<32 == 0)
		return 0;

	/* reduce x into [sqrt(2)/2, sqrt(2)] */
	hx += 0x3ff00000 - 0x3fe6a09e;
	k += (int)(hx>>20) - 0x3ff;
	hx = (hx&0x000fffff) + 0x3fe6a09e;
	u.i = (uint64_t)hx<<32 | (u.i&0xffffffff);
	x = u.f;

	f = x - 1.0;
	hfsq = 0.5*f*f;
	s = f/(2.0+f);
	z = s*s;
	w = z*z;
	t1 = w*(Lg2+w*(Lg4+w*Lg6));
	t2 = z*(Lg1+w*(Lg3+w*(Lg5+w*Lg7)));
	R = t2 + t1;

	/*
	 * f-hfsq must (for args near 1) be evaluated in extra precision
	 * to avoid a large cancellation when x is near sqrt(2) or 1/sqrt(2).
	 * This is fairly efficient since f-hfsq only depends on f, so can
	 * be evaluated in parallel with R.  Not combining hfsq with R also
	 * keeps R small (though not as small as a true `lo' term would be),
	 * so that extra precision is not needed for terms involving R.
	 *
	 * Compiler bugs involving extra precision used to break Dekker's
	 * theorem for spitting f-hfsq as hi+lo, unless double_t was used
	 * or the multi-precision calculations were avoided when double_t
	 * has extra precision.  These problems are now automatically
	 * avoided as a side effect of the optimization of combining the
	 * Dekker splitting step with the clear-low-bits step.
	 *
	 * y must (for args near sqrt(2) and 1/sqrt(2)) be added in extra
	 * precision to avoid a very large cancellation when x is very near
	 * these values.  Unlike the above cancellations, this problem is
	 * specific to base 2.  It is strange that adding +-1 is so much
	 * harder than adding +-ln2 or +-log10_2.
	 *
	 * This uses Dekker's theorem to normalize y+val_hi, so the
	 * compiler bugs are back in some configurations, sigh.  And I
	 * don't want to used double_t to avoid them, since that gives a
	 * pessimization and the support for avoiding the pessimization
	 * is not yet available.
	 *
	 * The multi-precision calculations for the multiplications are
	 * routine.
	 */

	/* hi+lo = f - hfsq + s*(hfsq+R) ~ log(1+f) */
	hi = f - hfsq;
	u.f = hi;
	u.i &= (uint64_t)-1<<32;
	hi = u.f;
	lo = f - hi - hfsq + s*(hfsq+R);

	val_hi = hi*ivln2hi;
	val_lo = (lo+hi)*ivln2lo + lo*ivln2hi;

	/* spadd(val_hi, val_lo, y), except for not using double_t: */
	y = k;
	w = y + val_hi;
	val_lo += (y - w) + val_hi;
	val_hi = w;

	return val_lo + val_hi;
}
