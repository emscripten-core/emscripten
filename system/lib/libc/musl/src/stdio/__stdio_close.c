#include "stdio_impl.h"
#include "aio_impl.h"

static int dummy(int fd)
{
	return fd;
}

weak_alias(dummy, __aio_close);

int __stdio_close(FILE *f)
{
#ifdef __EMSCRIPTEN__
	return __wasi_syscall_ret(__wasi_fd_close(__aio_close(f->fd)));
#else
	return syscall(SYS_close, __aio_close(f->fd));
#endif
}
