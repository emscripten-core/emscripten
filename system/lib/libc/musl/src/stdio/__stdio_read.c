#include "stdio_impl.h"
#include <sys/uio.h>

size_t __stdio_read(FILE *f, unsigned char *buf, size_t len)
{
	struct iovec iov[2] = {
		{ .iov_base = buf, .iov_len = len - !!f->buf_size },
		{ .iov_base = f->buf, .iov_len = f->buf_size }
	};
	ssize_t cnt;

#if __EMSCRIPTEN__
	size_t num;
	if (__wasi_syscall_ret(__wasi_fd_read(f->fd, (struct __wasi_iovec_t*)iov, 2, &num))) {
		num = -1;
	}
	cnt = num;
#else
	cnt = iov[0].iov_len ? syscall(SYS_readv, f->fd, iov, 2)
		: syscall(SYS_read, f->fd, iov[1].iov_base, iov[1].iov_len);
#endif
	if (cnt <= 0) {
		f->flags |= cnt ? F_ERR : F_EOF;
		return 0;
	}
	if (cnt <= iov[0].iov_len) return cnt;
	cnt -= iov[0].iov_len;
	f->rpos = f->buf;
	f->rend = f->buf + cnt;
	if (f->buf_size) buf[len-1] = *f->rpos++;
	return len;
}
