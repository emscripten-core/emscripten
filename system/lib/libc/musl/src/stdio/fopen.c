#include "stdio_impl.h"
#include <fcntl.h>
#include <string.h>
#include <errno.h>

FILE *fopen(const char *restrict filename, const char *restrict mode)
{
	FILE *f;
	int fd;
	int flags;

	/* Check for valid initial mode character */
	if (!strchr("rwa", *mode)) {
		errno = EINVAL;
		return 0;
	}

	/* Compute the flags to pass to open() */
	flags = __fmodeflags(mode);

	fd = sys_open(filename, flags, 0666);
	if (fd < 0) return 0;
#ifndef __EMSCRIPTEN__ // CLOEXEC makes no sense for a single process
	if (flags & O_CLOEXEC)
		__syscall(SYS_fcntl, fd, F_SETFD, FD_CLOEXEC);
#endif

	f = __fdopen(fd, mode);
	if (f) return f;

#ifdef __EMSCRIPTEN__
	__wasi_fd_close(fd);
#else
	__syscall(SYS_close, fd);
#endif
	return 0;
}
