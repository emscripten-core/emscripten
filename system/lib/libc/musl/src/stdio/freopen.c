#ifdef __wasilibc_unmodified_upstream // WASI has no syscall, dup
#include "stdio_impl.h"
#else
#include <wasi/api.h>
#endif
#include <fcntl.h>
#include <unistd.h>
#ifdef __wasilibc_unmodified_upstream // WASI has no syscall, dup
#else
// Move this below fcntl.h and unistd.h so that the __syscall macro doesn't
// cause trouble.
#include "stdio_impl.h"
#include <errno.h>
#endif

/* The basic idea of this implementation is to open a new FILE,
 * hack the necessary parts of the new FILE into the old one, then
 * close the new FILE. */

/* Locking IS necessary because another thread may provably hold the
 * lock, via flockfile or otherwise, when freopen is called, and in that
 * case, freopen cannot act until the lock is released. */

int __wasilibc_fd_renumber(int fd, int newfd) {
  __wasi_errno_t error = __wasi_fd_renumber(fd, newfd);
  if (error != 0) {
    errno = error;
    return -1;
  }
  return 0;
}

FILE *freopen(const char *restrict filename, const char *restrict mode, FILE *restrict f)
{
  int fl = __fmodeflags(mode);
  FILE *f2;

  FLOCK(f);

  fflush(f);

  if (!filename) {
    if (fl&O_CLOEXEC)
#ifdef __wasilibc_unmodified_upstream // WASI has no syscall
      __syscall(SYS_fcntl, f->fd, F_SETFD, FD_CLOEXEC);
#else
      fcntl(f->fd, F_SETFD, FD_CLOEXEC);
#endif
    fl &= ~(O_CREAT|O_EXCL|O_CLOEXEC);
#ifdef __wasilibc_unmodified_upstream // WASI has no syscall
    if (syscall(SYS_fcntl, f->fd, F_SETFL, fl) < 0)
#else
    if (fcntl(f->fd, F_SETFL, fl) < 0)
#endif
      goto fail;
  } else {
    f2 = fopen(filename, mode);
    if (!f2) goto fail;
    if (f2->fd == f->fd) f2->fd = -1; /* avoid closing in fclose */
#ifdef __wasilibc_unmodified_upstream // WASI has no dup
      else if (__dup3(f2->fd, f->fd, fl&O_CLOEXEC)<0) goto fail2;
#else
      // WASI doesn't have dup3, but does have a way to renumber
      // an existing file descriptor.
    else {
      if (__wasilibc_fd_renumber(f2->fd, f->fd)<0) goto fail2;
      f2->fd = -1; /* avoid closing in fclose */
    }
#endif

    f->flags = (f->flags & F_PERM) | f2->flags;
    f->read = f2->read;
    f->write = f2->write;
    f->seek = f2->seek;
    f->close = f2->close;

    fclose(f2);
  }

  FUNLOCK(f);
  return f;

  fail2:
  fclose(f2);
  fail:
  fclose(f);
  return NULL;
}

weak_alias(freopen, freopen64);
