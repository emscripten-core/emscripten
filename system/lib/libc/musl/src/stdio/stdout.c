#include "stdio_impl.h"

#if __EMSCRIPTEN__
// Emscripten doesn't support terminal seeking.
static off_t __emscripten_stdout_seek(FILE *f, off_t off, int whence)
{
	return 0;
}

// No special work is needed to close stdout.
static int __emscripten_stdout_close(FILE *f)
{
	return 0;
}
#endif

#undef stdout

static unsigned char buf[BUFSIZ+UNGET];
hidden FILE __stdout_FILE = {
	.buf = buf+UNGET,
	.buf_size = sizeof buf-UNGET,
	.fd = 1,
	.flags = F_PERM | F_NORD,
	.lbf = '\n',
#if __EMSCRIPTEN__
	// avoid stout_write which adds special terminal window size handling, which emscripten doesn't support anyhow
	.write = __stdio_write,
	.seek = __emscripten_stdout_seek,
	.close = __emscripten_stdout_close,
#else
	.write = __stdout_write,
	.seek = __stdio_seek,
	.close = __stdio_close,
#endif
	.lock = -1,
};
FILE *const stdout = &__stdout_FILE;
FILE *volatile __stdout_used = &__stdout_FILE;
