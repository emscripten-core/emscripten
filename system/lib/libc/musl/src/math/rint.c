#include <float.h>
#include <math.h>
#include <stdint.h>

#ifndef __wasm__
#if FLT_EVAL_METHOD==0 || FLT_EVAL_METHOD==1
#define EPS DBL_EPSILON
#elif FLT_EVAL_METHOD==2
#define EPS LDBL_EPSILON
#endif
static const double_t toint = 1/EPS;
#endif

double rint(double x)
{
// XXX EMSCRIPTEN: use the wasm instruction via clang builtin
// See https://github.com/emscripten-core/emscripten/issues/9236
#ifdef __wasm__
	return __builtin_rint(x);
#else
	union {double f; uint64_t i;} u = {x};
	int e = u.i>>52 & 0x7ff;
	int s = u.i>>63;
	double_t y;

	if (e >= 0x3ff+52)
		return x;
	if (s)
		y = x - toint + toint;
	else
		y = x + toint - toint;
	if (y == 0)
		return s ? -0.0 : 0;
	return y;
#endif
}
