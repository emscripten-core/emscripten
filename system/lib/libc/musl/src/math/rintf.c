#include <float.h>
#include <math.h>
#include <stdint.h>

#ifndef __wasm__
#if FLT_EVAL_METHOD==0
#define EPS FLT_EPSILON
#elif FLT_EVAL_METHOD==1
#define EPS DBL_EPSILON
#elif FLT_EVAL_METHOD==2
#define EPS LDBL_EPSILON
#endif
static const float_t toint = 1/EPS;
#endif

float rintf(float x)
{
// XXX EMSCRIPTEN: use the wasm instruction via clang builtin
// See https://github.com/emscripten-core/emscripten/issues/9236
#ifdef __wasm__
	return __builtin_rintf(x);
#else
	union {float f; uint32_t i;} u = {x};
	int e = u.i>>23 & 0xff;
	int s = u.i>>31;
	float_t y;

	if (e >= 0x7f+23)
		return x;
	if (s)
		y = x - toint + toint;
	else
		y = x + toint - toint;
	if (y == 0)
		return s ? -0.0f : 0.0f;
	return y;
#endif
}
