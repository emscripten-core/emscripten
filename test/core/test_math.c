/*
 * Copyright 2016 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

// For sincos
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char **argv) {
  printf("PI:%.2f,%.2f,%d\n", M_PI, -M_PI, (1 / 0.0) > 1e300);  // could end up as
                                                                // infinity, or just
                                                                // a very very big
                                                                // number
  printf("inf:");
  printf(",%d", isfinite(NAN) != 0);
  printf(",%d", isfinite(INFINITY) != 0);
  printf(",%d", isfinite(-INFINITY) != 0);
  printf(",%d", isfinite(12.3) != 0);
  printf(",%d", isinf(NAN) != 0);
  printf(",%d", isinf(INFINITY) != 0);
  printf(",%d", isinf(-INFINITY) != 0);
  printf(",%d", isinf(12.3) != 0);
  printf("\n");

  printf("div:");
  div_t div_result = div(23, 10);
  printf(",%d", div_result.quot);
  printf(",%d", div_result.rem);

  div_result = div(-5, 3); // see div manpage :)
  printf(",%d", div_result.quot); // -1
  printf(",%d", div_result.rem); // -2

  div_result = div(5, -3);
  printf(",%d", div_result.quot);
  printf(",%d", div_result.rem);

  div_result = div(-5, -3);
  printf(",%d", div_result.quot);
  printf(",%d", div_result.rem);
  printf("\n");

  {
    printf("modf:");
    double x,y;
    x = modf(3.2, &y);
    printf(",%1.1lf", x);
    printf(",%1.1lf", y);
    x = modf(-3.2, &y);
    printf(",%1.1lf", x);
    printf(",%1.1lf", y);
    printf("\n");
  }

  {
    printf("ldexp:");
    printf(",%1.1f", ldexp(3.0f, 3));
    printf(",%1.1lf", ldexpf(9.0, 5));
    printf(",%1.1Lf", ldexpl(10.0l, 2));
    printf("\n");
  }

  {
    printf("modff:");
    float x,y;
    x = modff(3.2, &y);
    printf(",%1.1f", x);
    printf(",%1.1f", y);
    x = modff(-3.2, &y);
    printf(",%1.1f", x);
    printf(",%1.1f", y);
    printf("\n");
  }

  printf("trig:");
  double sine = -1.0, cosine = -1.0;
  sincos(0.0, &sine, &cosine);
  printf(",%1.1lf", sine);
  printf(",%1.1lf", cosine);
  float fsine = -1.0f, fcosine = -1.0f;
  sincosf(0.0, &fsine, &fcosine);
  printf(",%1.1f", fsine);
  printf(",%1.1f", fcosine);
  fsine = sinf(1.1 + argc - 1);
  fcosine = cosf(1.1 + argc - 1);
  printf(",%1.1f", fsine);
  printf(",%1.1f", fcosine);
  printf("\n");
  return 0;
}
