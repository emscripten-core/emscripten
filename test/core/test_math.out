PI:3.14,-3.14,1
inf:,0,0,0,1,0,1,1,0
div:,2,3,-1,-2,-1,2,1,-2
modf:,0.2,3.0,-0.2,-3.0
ldexp:,24.0,288.0,40.0
modff:,0.2,3.0,-0.2,-3.0
trig:,0.0,1.0,0.0,1.0,0.9,0.5
